day8/alienalphabet.py

```diff
@@ -0,0 +1,131 @@
+##############
+## Solution ##
+##############
+
+def alienAlphabet(orderedWords):
+    # edge cases
+    if len(orderedWords) <= 0:
+        return ''
+    if len(orderedWords) == 1:
+        return removeDups(orderedWords[0])
+
+    # linearization of a DAG
+    graph = createDependencyGraph(orderedWords)
+    alphabet = topSort(graph)
+
+    return "".join(alphabet)
+
+#############
+## Helpers ##
+#############
+
+def createDependencyGraph(orderedWords):
+    graph = {}
+    for i, word in enumerate(orderedWords[:-1]):
+        nextWord = orderedWords[i+1]
+        shorter = min(len(word), len(nextWord))
+        for j in xrange(shorter):
+            letter = word[j]
+            otherLetter = nextWord[j]
+
+            if letter not in graph:
+                graph[letter] = []
+            if otherLetter not in graph:
+                graph[otherLetter] = []
+
+            if letter != otherLetter:
+                graph[letter].append(otherLetter)
+                break
+
+    return graph
+
+# graph is represented as key:value pair
+# key = node
+# value = [list, of, connected, nodes]
+#
+# the graph is also given to be a DAG so all the connected
+# nodes are from outgoing edges
+#
+# this function returns the resulting linearization in a list
+def topSort(graph):
+    sources = getSources(graph)
```

Comment:
i think a source here is defined to be a node with no incoming edges. then yes because such nodes would have an equal ordering in a linearization

cool observation